### PR TITLE
Implement @namespace for pages

### PR DESCRIPTION
**Pull request by rynowak**
This change adds support for @namespace for Pages, and introduces a set of
changes that are needed to support @namespace in the parser.

@namespace and @class have always been treated as reserved words by Razor,
with the intent that someday they would be allowed as directives.

This changes makes that possible.

You will still get an error about @namespace being a reserved word if you
don't have the directive.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Evolution/Legacy/CSharpCodeParser.cs [hunk 1]
```diff
@@ -110,6 +110,17 @@ protected void MapDirectives(Action handler, params string[] directives)
             {
                 _directiveParsers.Add(directive, handler);
                 Keywords.Add(directive);
+
+                // These C# keywords are reserved for use in directives. It's an error to use them outside of
+                // a directive. This code removes the error generation if the directive *is* registered.
+                if (string.Equals(directive, "class", StringComparison.OrdinalIgnoreCase))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/Legacy/CSharpCodeParser.cs]: This change was not a straightforward as Taylor said. Remove these as 'reserved words' unconditionally broke about 200 tests.

File: src/Microsoft.AspNetCore.Razor.Evolution/Legacy/CSharpCodeParser.cs [hunk 2]
```diff
@@ -295,8 +305,17 @@ private void AfterTransition()
                         }
                         else if (CurrentSymbol.Type == CSharpSymbolType.Keyword)
                         {
-                            KeywordBlock(topLevel: true);
-                            return;
+                            if (TryGetDirectiveHandler(CurrentSymbol.Content, out var handler))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/Legacy/CSharpCodeParser.cs]: These directives will be tokenized as keywords, so you need to handle that case.

**Comment by rynowak** [comment]
FYI this is blocked on #1194

It looks like this would cause a big diff if we fix it, so I want to pull it out separately.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+            // Includes the separator
+            return path.Substring(0, lastSeparator + 1);
+        }
+
+        private class NamespaceDirectiveNodeVisitor : RazorIRNodeWalker
+        {
+            // We want the last one, so get them all and then .
+            public DirectiveIRNode LastNamespaceDirective { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs]: `private set;` here and elsewhere.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+            var baseNamespace = directive.Tokens.First().Content;
+            if (string.IsNullOrEmpty(source) || directiveSource == null)
+            {
+                // No sources, can't compute a suffix.
+                @namespace = baseNamespace;
+                return false;
+            }
+
+            if (!source.StartsWith(directiveSource) || source.Length <= directiveSource.Length)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs]: ignorecase when comparing file paths.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            }
+
+            // OK so that this point we know that the 'imports' file containing this directive is in the directory
+            // hierarchy of this soure file. This is the case where we can append a suffix to the baseNamespace.
+            //
+            // Everything so far has just been defensiveness on our part.
+
+            var builder = new StringBuilder(baseNamespace);
+
+            var segments = source.Substring(directiveSource.Length).Split('/', '\\');
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs]: `.Split(Separators)`

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            if (lastSeparator == -1)
+            {
+                return null;
+            }
+
+            // Includes the separator
+            return path.Substring(0, lastSeparator + 1);
+        }
+
+        private class NamespaceDirectiveNodeVisitor : RazorIRNodeWalker
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs]: Any reason to not have 1 visitor that extracts the directives, namespaces and classes in a single pass?

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs]: Will probably need a test for the case sensitivity as well.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/NamespaceDirective.cs]: Seems fine

**Comment by rynowak** [comment]
@NTaylorMullen - rebased and updated

**Pull request merged**